prototype_source/maskedtensor_adagrad.rst

```diff
@@ -0,0 +1,182 @@
+    # This means that we're only applying the gradient to parts of the state_sum
+    # for which it is specified. This even drives the point home a lot more that
+    # the passed gradient is not sparse, but masked.
+    std = state_sum.sparse_mask(grad)
+
+    # We currently dodge all these concerns using the private method values.
+    std_values = std._values().sqrt_().add_(eps)
+
+    # We currently don't support div for sparse Tensors because zero / zero is
+    # not well defined. For a MaskedTensor undefined / undefined is undefined.
+    param.add_(_make_sparse(grad, grad_indices, grad_values / std_values), alpha=-clr)
+
+Line 24 is where we have a very important divergence.
+
+The addition of eps should technically be applied to all values but instead is only applied to specified values.
+Here we're using sparsity as a semantic extension and to enforce a certain pattern of defined and undefined values.
+If parts of the values of the gradient are zero, they are still included if materialized even though they
+could be compressed by other sparse storage layouts.
+This is technically quite brittle even though someone could argue that eps is always very small.
+
+Moreover, an implementation `add_` for sparsity as a storage layout and compression scheme should cause densification,
+but we force it not to for performance.
+For this one-off case it is fine.. until we want to introduce new compression schemes
+such as CSR, BSR, or 2:4 block sparsity. We'll then need to introduce separate Tensor types for each
+and write variations for gradients compressed using different storage formats, which is inconvenient.
+
+MaskedTensor sparse implementation
+----------------------------------
+
+We've been conflating sparsity as an optimization with sparsity as a semantic extension to PyTorch.
+MaskedTensor proposes to call the semantic extension through sparsity masked. 
+Currently we can't have dense semantics with sparse storage or masked semantics with dense storage, while 
+MaskedTensor fixes that because it separates the storage from the semantics.
+
+Consider the above example using a masked gradient:
+
+.. code-block:: python
+
+    # Create an entirely new set of parameters to avoid errors
+    param2 = torch.arange(8).reshape(2, 4).float()
+    state_sum2 = torch.full_like(param, 0.5)  # initial value for state sum
+
+    mask = (grad.to_dense() != 0).to_sparse()
+    masked_grad = masked_tensor(grad, mask)
+
+    state_sum2 = state_sum2 + masked_grad.pow(2).get_data()
+    std2 = masked_tensor(state_sum2.to_sparse(), mask)
+
+    # We can add support for in-place operations later. Notice how this doesn't
+    # need to access any storage internals and is in general a lot shorter
+    std2 = std2.sqrt().add(eps)
+
+    param2 = param2.add((masked_grad / std2).data(), alpha=-clr)
+
+Note that the implementations look quite similar but just shorter. Much of the boilerplate code
```

Comment:
```suggestion
Note that the implementations look quite similar but the MaskedTensor implementation is just shorter. Much of the boilerplate code
```

Seeing that _make_sparse is gone confirms my suspicions on line 68, and I do think it'd be helpful to highlight the points earlier so that the reader knows what to expect and has that "aha" moment while reading the new impl.